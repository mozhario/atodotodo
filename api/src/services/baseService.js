class BaseService {
  constructor(model) {
    this.model = model;
  }

  async getAll() {
    return this.model.find().sort({ createdAt: -1 });
  }

  async getById(id) {
    const item = await this.model.findById(id);
    if (!item) {
      throw new Error(`${this.model.modelName} not found`);
    }
    return item;
  }

  async create(data) {
    const item = new this.model(data);
    return item.save();
  }

  async update(id, updates) {
    const item = await this.getById(id);
    Object.assign(item, updates);
    return item.save();
  }

  async delete(id) {
    const result = await this.model.findByIdAndDelete(id);
    if (!result) {
      throw new Error(`${this.model.modelName} not found`);
    }
    return result;
  }
}

export default BaseService; 