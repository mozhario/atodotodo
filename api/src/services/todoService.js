import BaseService from './baseService.js';
import Todo from '../models/Todo.js';

class TodoService extends BaseService {
  constructor() {
    super(Todo);
  }

  async toggleComplete(id) {
    const todo = await this.getById(id);
    todo.completed = !todo.completed;
    return todo.save();
  }
}

export default new TodoService(); 