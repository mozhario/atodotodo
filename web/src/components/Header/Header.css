.header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 2rem;
}

.header h1 {
  margin: 0;
}

.header-controls {
  display: flex;
  gap: 0.5rem;
}

.btn-icon {
  background: none;
  border: none;
  font-size: 1.5rem;
  cursor: pointer;
  padding: 0.25rem;
  display: flex;
  align-items: center;
  justify-content: center;
  transition: transform 0.2s;
}

.btn-icon:hover {
  transform: scale(1.1);
} 