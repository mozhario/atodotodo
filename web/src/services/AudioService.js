import { AUDIO_CONFIG } from '../config/audio';

class AudioService {
  constructor() {
    this.sounds = {
      add: [],
      delete: [],
      check: [],
      uncheck: []
    };
    this.currentAudio = null;
    this.currentTimer = null;
    const storedValue = localStorage.getItem(AUDIO_CONFIG.AUDIO_ENABLED_STORAGE_KEY);
    this.enabled = storedValue === null ? AUDIO_CONFIG.DEFAULT_AUDIO_ENABLED : storedValue !== 'false';
    this.loadSounds();
  }

  setEnabled(enabled) {
    this.enabled = enabled;
    localStorage.setItem(AUDIO_CONFIG.AUDIO_ENABLED_STORAGE_KEY, enabled);
    if (!enabled) {
      this.stopCurrentAudio();
    }
  }

  isEnabled() {
    return this.enabled;
  }

  loadSounds() {
    AUDIO_CONFIG.CATEGORIES.forEach(category => {
      const files = Object.keys(
        import.meta.glob('/public/audio/todo/*/*.{mp3,ogg,wav}', { 
          as: 'url',
          eager: true 
        })
      );
      
      this.sounds[category] = files
        .filter(path => path.includes(`/todo/${category}/`))
        .map(path => path.replace('/public', ''));
    });
  }

  getRandomSound(category) {
    const sounds = this.sounds[category];
    if (!sounds || sounds.length === 0) return null;
    
    const randomIndex = Math.floor(Math.random() * sounds.length);
    return sounds[randomIndex];
  }

  stopCurrentAudio() {
    if (this.currentAudio) {
      this.currentAudio.pause();
      this.currentAudio.currentTime = 0;
      this.currentAudio = null;
    }
    if (this.currentTimer) {
      clearTimeout(this.currentTimer);
      this.currentTimer = null;
    }
  }

  async playSound(category) {
    if (!this.enabled) return;

    const soundPath = this.getRandomSound(category);
    if (!soundPath) return;

    try {
      this.stopCurrentAudio();

      const audio = new Audio(soundPath);
      this.currentAudio = audio;
      audio.volume = AUDIO_CONFIG.VOLUME;
      
      this.currentTimer = setTimeout(() => {
        this.stopCurrentAudio();
      }, AUDIO_CONFIG.DURATION * 1000);

      audio.addEventListener('ended', () => {
        if (this.currentTimer) {
          clearTimeout(this.currentTimer);
          this.currentTimer = null;
        }
        this.currentAudio = null;
      });

      await audio.play();
    } catch (error) {
      console.error('Failed to play sound:', error);
    }
  }

  playAddSound() {
    return this.playSound('add');
  }

  playDeleteSound() {
    return this.playSound('delete');
  }

  playCheckSound() {
    return this.playSound('check');
  }

  playUncheckSound() {
    return this.playSound('uncheck');
  }
}

export const audioService = new AudioService(); 