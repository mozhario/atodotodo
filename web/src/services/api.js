import axios from 'axios';

const api = axios.create({
  baseURL: '/api',
});

// Add a request interceptor to add the auth token to all requests
api.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('token');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// Add a response interceptor to handle 401 errors
api.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response?.status === 401) {
      localStorage.removeItem('token');
      window.location.reload();
    }
    return Promise.reject(error);
  }
);

export default api; 