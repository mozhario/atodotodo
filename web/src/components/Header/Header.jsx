import './Header.css';

function Header({ onLogout }) {
  return (
    <div className="header">
      <h1 className="text-center mb-4">a todo to do</h1>
      <button onClick={onLogout} className="btn btn-danger">
        Logout
      </button>
    </div>
  );
}

export default Header; 