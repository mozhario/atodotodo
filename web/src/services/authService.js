import api from './api';

export const authService = {
  async signIn(username, password) {
    const response = await api.post('/auth/login', { username, password });
    localStorage.setItem('token', response.data.token);
    return response.data;
  },

  async signUp(username, password) {
    const response = await api.post('/auth/register', { username, password });
    localStorage.setItem('token', response.data.token);
    return response.data;
  },

  signOut() {
    localStorage.removeItem('token');
  },

  isSignedIn() {
    return !!localStorage.getItem('token');
  }
}; 