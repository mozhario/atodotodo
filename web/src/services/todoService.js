import api from './api';

export const todoService = {
  async getUserTodos() {
    const response = await api.get('/todos');
    return response.data;
  },

  async createTodo(title) {
    const response = await api.post('/todos', { title });
    return response.data;
  },

  async toggleTodoCompletion(id) {
    const response = await api.post(`/todos/${id}/toggle`);
    return response.data;
  },

  async updateTodoTitle(id, title) {
    const response = await api.patch(`/todos/${id}`, { title });
    return response.data;
  },

  async removeTodo(id) {
    await api.delete(`/todos/${id}`);
    return id;
  }
}; 