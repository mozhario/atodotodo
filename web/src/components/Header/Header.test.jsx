import { render, screen, fireEvent } from '@testing-library/react';
import { describe, test, expect, vi } from 'vitest';
import Header from './Header';
import { audioService } from '../../services/AudioService';

// Mock the audioService
vi.mock('../../services/AudioService', () => ({
  audioService: {
    isEnabled: vi.fn(() => true),
    setEnabled: vi.fn(),
  },
}));

describe('Header Component', () => {
  test('renders the header with title and buttons', () => {
    const mockLogout = vi.fn();
    const { container } = render(<Header onLogout={mockLogout} />);

    // Check if the title is rendered
    expect(screen.getByText('a todo to do')).toBeInTheDocument();

    // Check if the audio toggle button is rendered
    const audioButton = screen.getByTitle('Disable sounds');
    expect(audioButton).toBeInTheDocument();

    // Check if the logout button is rendered
    const logoutButton = screen.getByTitle('Logout');
    expect(logoutButton).toBeInTheDocument();

    // Snapshot test
    expect(container).toMatchSnapshot();
  });

  test('toggles audio state when audio button is clicked', () => {
    const mockLogout = vi.fn();
    render(<Header onLogout={mockLogout} />);

    const audioButton = screen.getByTitle('Disable sounds');
    fireEvent.click(audioButton);

    // Check if the audioService.setEnabled was called
    expect(audioService.setEnabled).toHaveBeenCalledWith(false);
  });

  test('calls onLogout when logout button is clicked', () => {
    const mockLogout = vi.fn();
    render(<Header onLogout={mockLogout} />);

    const logoutButton = screen.getByTitle('Logout');
    fireEvent.click(logoutButton);

    // Check if the onLogout function was called
    expect(mockLogout).toHaveBeenCalled();
  });
}); 