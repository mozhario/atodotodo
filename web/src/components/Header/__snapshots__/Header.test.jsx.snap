// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header Component > renders the header with title and buttons 1`] = `
<div>
  <header
    class="header"
  >
    <h1>
      a todo to do
    </h1>
    <div
      class="header-controls"
    >
      <button
        class="btn-icon"
        title="Disable sounds"
      >
        🔉
      </button>
      <button
        class="btn-icon"
        title="Logout"
      >
        ❌
      </button>
    </div>
  </header>
</div>
`;
